mempool/iterators.go

```diff
@@ -0,0 +1,208 @@
+package mempool
+
+import (
+	"fmt"
+
+	"github.com/cometbft/cometbft/internal/clist"
+	"github.com/cometbft/cometbft/types"
+)
+
+// WRRIterator is the base struct for implementing iterators that traverse lanes with
+// the classical Weighted Round Robin (WRR) algorithm.
+type WRRIterator struct {
+	sortedLanes []types.Lane
+	laneIndex   int                            // current lane being iterated; index on sortedLanes
+	counters    map[types.Lane]uint            // counters of consumed entries, for WRR algorithm
+	cursors     map[types.Lane]*clist.CElement // last accessed entries on each lane
+}
+
+func (iter *WRRIterator) nextLane() types.Lane {
+	iter.laneIndex = (iter.laneIndex + 1) % len(iter.sortedLanes)
+	return iter.sortedLanes[iter.laneIndex]
+}
+
+// Non-blocking version of the WRR iterator to be used for reaping and
+// rechecking transactions.
+//
+// Lock must be held on the mempool when iterating: the mempool cannot be
+// modified while iterating.
+type NonBlockingWRRIterator struct {
+	WRRIterator
+}
+
+func NewWRRIterator(mem *CListMempool) *NonBlockingWRRIterator {
+	baseIter := WRRIterator{
+		sortedLanes: mem.sortedLanes,
+		counters:    make(map[types.Lane]uint, len(mem.lanes)),
+		cursors:     make(map[types.Lane]*clist.CElement, len(mem.lanes)),
+	}
+	iter := &NonBlockingWRRIterator{
+		WRRIterator: baseIter,
+	}
+	iter.Reset(mem.lanes)
+	return iter
+}
+
+// Reset must be called before every use of the iterator.
+func (iter *NonBlockingWRRIterator) Reset(lanes map[types.Lane]*clist.CList) {
```

Comment:
I see that after fixing the possible race condition, this is only called from `NewWRRIterator` above, and ReCheck.

Would it make sense to make this method private, and have ReCheck call `NewWRRIterator` rather than `Reset`? This way we reduce the possibility of future misuse.

Reply:
I implemented this change you propose in #4013. I call `NewNonBlockingIterator` rather than `Reset` when rechecking, and then remove `Reset`.